File: heaps/2163-minimum-difference-in-sums-after-removal-of-elements.py
import heapq
from typing import List
import unittest

# https://leetcode.com/problems/minimum-difference-in-sums-after-removal-of-elements/

# python3 -m unittest heaps/2163-minimum-difference-in-sums-after-removal-of-elements.py


class Solution(unittest.TestCase):
    # Time: O(n log n)
    # Space: O(n)
    def minimumDifference(self, nums: List[int]) -> int:
        n = len(nums) // 3
        sumsec = [0] * (n + 1)
        sumsec[n] = sum(nums[2 * n :])
        mnhp = nums[2 * n :][:]
        heapq.heapify(mnhp)
        for i in range(n - 1, -1, -1):
            heapq.heappush(mnhp, nums[n + i])
            sumsec[i] = sumsec[i + 1] + nums[n + i] - heapq.heappop(mnhp)
        sumfir = sum(nums[:n])
        mxhp = [-x for x in nums[:n]]
        heapq.heapify(mxhp)
        diff = sumfir - sumsec[0]
        for i in range(n):
            heapq.heappush(mxhp, -nums[n + i])
            sumfir += nums[n + i] + heapq.heappop(mxhp)
            diff = min(diff, sumfir - sumsec[i + 1])
        return diff

    def test(self):
        for nums, expected in [
            ([3, 1, 2], -1),
            ([7, 9, 5, 8, 1, 3], 1),
            ([7, 9, 5, 8, 1, 3, 1, 1, 10000], -9993),
            ([16, 46, 43, 41, 42, 14, 36, 49, 50, 28, 38, 25, 17, 5, 18, 11, 14, 21, 23, 39, 23], -14),
        ]:
            output = self.minimumDifference(nums)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: heaps/2163-minimum-difference-in-sums-after-removal-of-elements.go
package heaps

import (
	"container/heap"

	"github.com/shahzodshafizod/leetcode/pkg"
)

// https://leetcode.com/problems/minimum-difference-in-sums-after-removal-of-elements/

// Time: O(n log n)
// Space: O(n)
func minimumDifference(nums []int) int64 {
	n := len(nums) / 3
	mnhp := pkg.NewHeap(make([]int, 0), func(x, y int) bool { return x < y })
	sumsec := make([]int64, n+1)

	for i := n - 1; i >= 0; i-- {
		heap.Push(mnhp, nums[2*n+i])
		sumsec[n] += int64(nums[2*n+i])
	}

	for i := n - 1; i >= 0; i-- {
		heap.Push(mnhp, nums[n+i])
		sumsec[i] = sumsec[i+1] + int64(nums[n+i]) - int64(heap.Pop(mnhp).(int))
	}

	mxhp := pkg.NewHeap(make([]int, 0), func(x, y int) bool { return x > y })

	var sumfir int64 = 0

	for i := 0; i < n; i++ {
		heap.Push(mxhp, nums[i])
		sumfir += int64(nums[i])
	}

	var diff int64 = sumfir - sumsec[0]

	for i := 0; i < n; i++ {
		heap.Push(mxhp, nums[n+i])
		sumfir += int64(nums[n+i]) - int64(heap.Pop(mxhp).(int))
		diff = min(diff, sumfir-sumsec[i+1])
	}

	return diff
}
